File: csharp/cli/src/Loader.cs
using Ball.Engine;

namespace Ball.Cli;

/// <summary>
/// Loading a <c>.ball.json</c>/<c>.ball.bin</c> program for every subcommand (issue #385).
///
/// <para>Reuses <see cref="Ball.Engine.BallEngine"/>'s own loader
/// (<see cref="Ball.Engine.BallEngine.FromJson"/>/<see cref="Ball.Engine.BallEngine.FromBinary"/>,
/// backed by <c>csharp/engine/src/Loader.cs</c>'s proto3-JSON&lt;-&gt;binary round trip) rather
/// than re-implementing program parsing here — the C# analog of
/// <c>rust/cli/src/loader.rs</c>.</para>
///
/// <para>Format is sniffed by extension: a path ending in <c>.bin</c> is read as raw bytes and
/// decoded as binary protobuf; anything else (<c>.ball.json</c>, <c>.json</c>, or no extension at
/// all) is read as UTF-8 text and decoded as proto3 JSON (the <c>@type</c> Any envelope, if
/// present, is stripped by the engine loader). Mirrors
/// <c>dart/cli/lib/src/runner.dart</c>'s <c>path.endsWith('.bin')</c> convention and the
/// Rust/TS CLIs' own format-sniffing rule.</para>
/// </summary>
public static class Loader
{
    /// <summary>
    /// Load a target <see cref="Ball.Engine.BallEngine"/> from <paramref name="path"/>. I/O
    /// failures (missing file, permission error, ...) become <see cref="CliIoError"/> (exit
    /// <c>3</c>); a malformed/undecodable program becomes <see cref="CliParseError"/> (exit
    /// <c>2</c>) — see the type doc comment for the format-sniffing rule.
    /// </summary>
    public static BallEngine LoadEngine(string path)
    {
        var isBinary = path.EndsWith(".bin", StringComparison.Ordinal);
        if (isBinary)
        {
            byte[] bytes;
            try
            {
                bytes = File.ReadAllBytes(path);
            }
            catch (Exception e) when (e is IOException or UnauthorizedAccessException)
            {
                throw new CliIoError($"could not read {path}: {e.Message}");
            }

            try
            {
                return BallEngine.FromBinary(bytes);
            }
            catch (EngineException e)
            {
                throw new CliParseError(e.Message);
            }
        }

        string text;
        try
        {
            text = File.ReadAllText(path);
        }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException)
        {
            throw new CliIoError($"could not read {path}: {e.Message}");
        }

        try
        {
            return BallEngine.FromJson(text);
        }
        catch (EngineException e)
        {
            throw new CliParseError(e.Message);
        }
    }
}

File: rust/cli/src/loader.rs
//! Loading a `.ball.json`/`.ball.bin` program for every subcommand (issue
//! #41).
//!
//! Reuses `ball-lang-engine`'s own loader (`BallEngine::from_json`/`from_binary`,
//! backed by `rust/engine/src/loader.rs`'s prost-reflect proto3-JSON<->binary
//! round trip) rather than re-implementing program parsing here — the same
//! codec every subcommand (`run`, `compile`, `check`) needs, and the one the
//! issue explicitly calls out ("the loader in ball-lang-engine handles both —
//! reuse it").
//!
//! Format is sniffed by extension: a path ending in `.bin` is read as raw
//! bytes and decoded as binary protobuf; anything else (`.ball.json`, `.json`,
//! or no extension at all) is read as UTF-8 text and decoded as proto3 JSON
//! (the `@type` Any envelope, if present, is stripped by the engine loader).
//! This mirrors `dart/cli/lib/src/runner.dart`'s `_loadBallFile`
//! (`path.endsWith('.bin')`) and the Dart/TS engines' own format convention.
use std::path::Path;

use ball_lang_engine::BallEngine;

use crate::error::CliError;

/// Load a target `Program` (wrapped in its [`BallEngine`], which the `run`
/// subcommand needs directly and `compile`/`check` read `.program()` off of)
/// from `path`. I/O failures (missing file, permission error, ...) become
/// [`CliError::Io`] (exit `3`); a malformed/undecodable program becomes
/// [`CliError::Parse`] (exit `2`) — see the module doc comment for the
/// format-sniffing rule.
pub fn load_engine(path: &Path) -> Result<BallEngine, CliError> {
    let is_binary = path.extension().and_then(|ext| ext.to_str()) == Some("bin");
    if is_binary {
        let bytes = std::fs::read(path)
            .map_err(|e| CliError::Io(format!("could not read {}: {e}", path.display())))?;
        BallEngine::from_binary(&bytes).map_err(CliError::from)
    } else {
        let text = std::fs::read_to_string(path)
            .map_err(|e| CliError::Io(format!("could not read {}: {e}", path.display())))?;
        BallEngine::from_json(&text).map_err(CliError::from)
    }
}
